repo-scripts/size-analysis/analysis-helper.ts

```diff
@@ -0,0 +1,779 @@
+ *
+ * Examples of Various Type of Exports
+ * FunctionDeclaration: export function aFunc(): string {...};
+ * ClassDeclaration: export class aClass {};
+ * EnumDeclaration: export enum aEnum {};
+ * VariableDeclaration: export let aVariable: string;
+ * VariableStatement: export const aVarStatement: string = "string"; export const { a, b } = { a: 'a', b: 'b' };
+ * ExportDeclaration:
+ *      named exports: export {foo, bar} from '...'; export {foo as foo1, bar} from '...'; export {LogLevel};
+ *      export everything: export * from '...';
+ */
+export function extractDeclarations(
+  filePath: string,
+  map?: Map<string, string>
+): MemberList {
+  const program = ts.createProgram([filePath], { allowJs: true });
+  const checker = program.getTypeChecker();
+
+  const sourceFile = program.getSourceFile(filePath);
+  if (!sourceFile) {
+    throw new Error(`${ErrorCode.FILE_PARSING_ERROR} ${filePath}`);
+  }
+
+  let declarations: MemberList = {
+    functions: [],
+    classes: [],
+    variables: [],
+    enums: [],
+    externals: []
+  };
+  // define a map here which is used to handle export statements like "export {LogLevel}".
+  // As there is no from clause in such export statements, we retrieve symbol location by parsing the corresponding import
+  // statements. We store the symbol and its defined location as key value pairs in the map.
+  const symbolLocation: Map<string, string> = new Map();
+
+  ts.forEachChild(sourceFile, node => {
+    if (ts.isFunctionDeclaration(node)) {
+      declarations.functions.push(node.name!.text);
+    } else if (ts.isClassDeclaration(node)) {
+      declarations.classes.push(node.name!.text);
+    } else if (ts.isVariableDeclaration(node)) {
+      declarations.variables.push(node.name!.getText());
+    } else if (ts.isEnumDeclaration(node)) {
+      declarations.enums.push(node.name.escapedText.toString());
+    } else if (ts.isVariableStatement(node)) {
+      const variableDeclarations = node.declarationList.declarations;
+
+      variableDeclarations.forEach(variableDeclaration => {
+        //variableDeclaration.name could be of Identifier type or of BindingPattern type
+        // Identifier Example: export const a: string = "aString";
+        if (ts.isIdentifier(variableDeclaration.name)) {
+          declarations.variables.push(
+            variableDeclaration.name.getText(sourceFile)
+          );
+        }
+        // Binding Pattern Example: export const {a, b} = {a: 1, b: 1};
+        else {
+          variableDeclaration.name.elements.forEach(node => {
+            declarations.variables.push(node.name.getText(sourceFile));
+          });
+        }
+      });
+    } else if (ts.isImportDeclaration(node) && node.importClause) {
+      const symbol = checker.getSymbolAtLocation(node.moduleSpecifier);
+      if (symbol && symbol.valueDeclaration) {
+        const importFilePath = symbol.valueDeclaration.getSourceFile().fileName;
+        // import { a, b } from '@firebase/dummy-exp'
```

Comment:
It doesn't seem we are handling renamed imports, e.g. `import { a as AA} from ...`. You may need to store both original name and renamed name in order to handle exports without imports correctly.